src/main/resources/frontend/src/network/ServerCommunicator.ts

```diff
@@ -0,0 +1,70 @@
+import { useAuthStore } from '@/stores/auth'
+import { ServerError } from '@/network/ServerError'
+import { useAppConfigStore } from '@/stores/appConfig'
+
+export const ServerCommunicator = {
+  getRequest: getRequest,
+  postRequest: postRequest,
+  patchRequest: patchRequest,
+  doUnprocessedRequest: doUnprocessedRequest
+}
+
+async function getRequest<T>(endpoint: string, expectResponse?: boolean): Promise<T> {
+  return await doRequest<T>("GET", endpoint, null,expectResponse ?? true)
+}
+
+async function postRequest<T>(endpoint: string, bodyObject?: Object, expectResponse?: boolean): Promise<T> {
+  return await doRequest<T>("POST", endpoint, bodyObject,expectResponse ?? true)
+}
+
+async function patchRequest<T>(endpoint: string, bodyObject?: Object, expectResponse?: boolean): Promise<T> {
+  return await doRequest<T>("PATCH", endpoint, bodyObject,expectResponse ?? true)
+}
+
+async function doRequest<T>(method: string,
+                            endpoint: string,
+                            bodyObject?: Object | null,
+                            expectResponse?: boolean): Promise<T> {
+  const response = await doUnprocessedRequest(method, endpoint, bodyObject)
+
+  if (!expectResponse) {
+    return null as T
+  }
+
+  // this makes sure there is actually something in the body before trying to parse
+  const text = await response.text()
+  if (text) {
+    return JSON.parse(text) as T
```

Comment:
As discussed in CS340, this move to interpret the data as `T` is convenient, but not safe. The backend could totally lie or break or change, or the front-end could be hitting the wrong endpoint, and the data could not actually line up with the type we expect it to be.

A less convenient but safer way to access this data would be by returning a type `unknown`. This would force users of the function to do an actual content check before typecasting it, which would not only give them an opportunity to fail more gracefully when it's wrong, but also be more confident that the data is actually correct.

It's not necessary to make this change, but I just want to make sure we're acknowledging the tradeoffs of this decision.

Reply:
If this is a concern, I'm pretty sure we can loop through the json object `.keys()` and make sure it has all the members, before returning, and make it just a part of the request methods.

Reply:
I totally agree. I'm glad we were able to at least discuss and be aware of the consequences of this decision.

This is the same problem Dr. Wilkerson warned us about in our CS 340 class last week, but it will be acceptable to move in this direction.